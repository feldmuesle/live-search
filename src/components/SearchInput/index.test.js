import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { render, fireEvent, screen, within } from '@testing-library/react';
import setupIcons from '../../setup-icons';
import { SearchInput } from '.';
import useStringSearch from '../../hooks/use-string-search';

setupIcons();

const data = [
  { name: 'cat', id: '1' },
  { name: 'catastrophe', id: '2' },
  { name: 'category', id: '3' },
  { name: 'tiger', id: '4' },
];

const getResult = useStringSearch(['name', 'name']);

const formatSelection = ({ name }) => `${name}`;

// eslint-disable-next-line react/prop-types
function ResultComponent({ name }) {
  return <div>{name}</div>;
}

describe('SearchInput renders correctly', () => {
  it('renders correctly when closed', () => {
    const component = renderer.create(
      <SearchInput
        data={data}
        getResult={getResult}
        resultComponent={ResultComponent}
        formatSelection={formatSelection}
      />
    );

    const snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
  });
  it('renders correctly when user clicks on input', () => {
    const component = renderer.create(
      <SearchInput
        data={data}
        getResult={getResult}
        resultComponent={ResultComponent}
        formatSelection={formatSelection}
      />
    );

    const input = component.root.findByType('input');
    act(() => {
      input.props.onFocus({ target: { setAttribute: () => {} } });
    });

    const snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});

describe('user interaction', () => {
  describe('when users focusses on empty input', () => {
    it('opens dropdown on focus', () => {
      render(
        <SearchInput
          data={data}
          getResult={getResult}
          resultComponent={ResultComponent}
          formatSelection={formatSelection}
        />
      );
      const input = screen.getByRole('textbox');
      input.focus();
      const dropdown = screen.getByRole('list');

      expect(dropdown).toBeTruthy();
    });

    it('it displays all items', () => {
      render(
        <SearchInput
          data={data}
          getResult={getResult}
          resultComponent={ResultComponent}
          formatSelection={formatSelection}
        />
      );
      const input = screen.getByRole('textbox');
      input.focus();
      const dropdown = screen.getByRole('list');
      const listitems = within(dropdown).getAllByRole('listitem');

      expect(listitems.length).toEqual(4);
    });

    it('displays correct search results for current value', () => {
      render(
        <SearchInput
          data={data}
          getResult={getResult}
          resultComponent={ResultComponent}
          formatSelection={formatSelection}
        />
      );
      const input = screen.getByRole('textbox');
      input.focus();

      const listItem = screen.getByText('cat');
      fireEvent.mouseDown(listItem);

      input.focus();
      const listItems = screen.queryAllByRole('listitem');

      expect(listItems.length).toEqual(3);
    });
  });

  describe('user types into input', () => {
    it('calls function for getting results', () => {
      const onChange = jest.fn(() => data);
      render(
        <SearchInput
          data={data}
          getResult={onChange}
          resultComponent={ResultComponent}
          formatSelection={formatSelection}
        />
      );
      const value = 'cat';
      const input = screen.getByRole('textbox');
      fireEvent.change(input, { target: { value } });

      expect(onChange).toHaveBeenCalled();
    });
  });

  describe('user clicks on item in dropdown', () => {
    it('calls function for formatting selection', () => {
      const format = jest.fn(() => 'something');
      render(
        <SearchInput
          data={data}
          getResult={getResult}
          resultComponent={ResultComponent}
          formatSelection={format}
        />
      );
      const input = screen.getByRole('textbox');
      input.focus();

      const selection = screen.getByText('cat');
      fireEvent.mouseDown(selection);

      expect(format).toHaveBeenCalled();
    });

    it('updates value with formatted selection and closes dropdown', () => {
      render(
        <SearchInput
          data={data}
          getResult={getResult}
          resultComponent={ResultComponent}
          formatSelection={formatSelection}
        />
      );
      const input = screen.getByRole('textbox');
      input.focus();

      const selection = screen.getByText('cat');
      fireEvent.mouseDown(selection);

      const dropdown = screen.queryByRole('list');

      expect(input.value).toBe('cat');
      expect(dropdown).toBeFalsy();
    });
  });

  describe('input looses focus', () => {
    it('closes dropdown', () => {
      render(
        <SearchInput
          data={data}
          getResult={getResult}
          resultComponent={ResultComponent}
          formatSelection={formatSelection}
        />
      );
      const input = screen.getByRole('textbox');
      input.focus();
      input.blur();

      const dropdown = screen.queryByRole('list');

      expect(dropdown).toBeFalsy();
    });

    it('keeps the current value', () => {
      const format = ({ name }) => `${name} ${name}`;
      render(
        <SearchInput
          data={data}
          getResult={getResult}
          resultComponent={ResultComponent}
          formatSelection={format}
        />
      );
      const input = screen.getByRole('textbox');
      input.focus();

      const listItem = screen.getByText('cat');
      fireEvent.mouseDown(listItem);

      input.focus();
      input.blur();

      expect(input.value).toBe('cat cat');
    });
  });
});
