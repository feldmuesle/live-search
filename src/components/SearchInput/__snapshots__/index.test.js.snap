// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput renders correctly renders correctly when closed 1`] = `
<div
  className="search-input"
>
  <div
    className="form-item"
  >
    <label
      className="form-item__label"
    >
      Manager:
      <input
        autoComplete="new-password"
        className="input"
        name="managers"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Choose Manager"
        required={false}
        type="text"
        value=""
      />
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-chevron-up fa-w-14 search-input__icon"
        data-icon="chevron-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </label>
  </div>
  <div
    className="search-input__dropdown"
  >
    <ul
      className="search-input__results"
    >
      <li
        className="search-input__result-item"
        onMouseDown={[Function]}
      >
        <div>
          cat
        </div>
      </li>
      <li
        className="search-input__result-item"
        onMouseDown={[Function]}
      >
        <div>
          catastrophe
        </div>
      </li>
      <li
        className="search-input__result-item"
        onMouseDown={[Function]}
      >
        <div>
          category
        </div>
      </li>
      <li
        className="search-input__result-item"
        onMouseDown={[Function]}
      >
        <div>
          tiger
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`SearchInput renders correctly renders correctly when user clicks on input 1`] = `
<div
  className="search-input"
>
  <div
    className="form-item"
  >
    <label
      className="form-item__label"
    >
      Manager:
      <input
        autoComplete="new-password"
        className="input"
        name="managers"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Choose Manager"
        required={false}
        type="text"
        value=""
      />
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-chevron-up fa-w-14 search-input__icon"
        data-icon="chevron-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </label>
  </div>
  <div
    className="search-input__dropdown"
  >
    <ul
      className="search-input__results"
    >
      <li
        className="search-input__result-item"
        onMouseDown={[Function]}
      >
        <div>
          cat
        </div>
      </li>
      <li
        className="search-input__result-item"
        onMouseDown={[Function]}
      >
        <div>
          catastrophe
        </div>
      </li>
      <li
        className="search-input__result-item"
        onMouseDown={[Function]}
      >
        <div>
          category
        </div>
      </li>
      <li
        className="search-input__result-item"
        onMouseDown={[Function]}
      >
        <div>
          tiger
        </div>
      </li>
    </ul>
  </div>
</div>
`;
